### PR TITLE
Remove the manifest schema conversion machinery

### PR DESCRIPTION
**Pull request by git-hyagi**
closes: #1509

### CODE REVIEW

File: unittest_requirements.txt [hunk 1]
```diff
@@ -2,3 +2,4 @@ mock
 pytest-django
 pulp-smash @ git+https://github.com/pulp/pulp-smash.git
 pytest<8
+pyjwkest>=1.4,<=1.4.2
```
Comment (lubosmj) [inline unittest_requirements.txt:5]: Do we still need this dependency?

File: pulp_container/tests/functional/api/test_content_cache.py [hunk 2]
```diff
@@ -186,6 +192,11 @@ def fetch_response_metadata(self, response):
         if DEFAULT_FILE_STORAGE in STANDARD_FILE_STORAGE_FRAMEWORKS:
             return response.status_code, response.headers.get("X-PULP-CACHE")
         else:
+            """Corner case for deprecated v2 schema1"""
```
Comment (lubosmj) [inline pulp_container/tests/functional/api/test_content_cache.py:195]: Which corner cases are we catching here? I thought that we were dealing only with manifest v2 schema 2 content inside this case.

File: pulp_container/app/registry.py [hunk 3]
```diff
@@ -231,6 +229,11 @@ async def dispatch_tag(self, request, tag, response_headers):
                 streamed back to the client.
 
         """
+        # return 404 in case the client is requesting docker manifest v2 schema 1
```
Comment (lubosmj) [inline pulp_container/app/registry.py:232]: I think this part is not required inside `dispatch_tag`. How about we remove the code and add the highlighted comment just above the `raise PathNotResolved` call placed at the end of the `get_tag` method? So, we will not forget about the rationale behind raising the exception.

File: unittest_requirements.txt (hunk 1 above)
Comment (git-hyagi) [reply unittest_requirements.txt:5]: Hum... I probably had issues before removing the test_convert.py file:
https://github.com/pulp/pulp_container/pull/1529/files#diff-2f5092e427a2a7997e2ce75c8c252cbd9d3d30b33f2ab32dbe1b241194de98bdL6
I just tried again without it and had no dependency errors. Removing it.

File: pulp_container/tests/functional/api/test_content_cache.py (hunk 2 above)
Comment (git-hyagi) [reply pulp_container/tests/functional/api/test_content_cache.py:195]: I don't remember now, but it was probably an error in test_05_different_headers.
I'm rerunning the tests without this block to confirm that the "corner case" was actually because of wrong headers in my tests.

File: pulp_container/app/registry.py (hunk 3 above)
Comment (git-hyagi) [reply pulp_container/app/registry.py:232]: > I think this part is not required inside dispatch_tag.

Agreed. I added this part when I was having issues with the test_05_different_headers and setting "_wrong headers_". I'm rerunning the tests without this modification in dispatch_tag and will check if the tests will pass.

> How about we remove the code and add the highlighted comment just above the raise PathNotResolved call placed at the end of the get_tag method? So, we will not forget about the rationale behind raising the exception.

Good idea!

**Comment by ipanova** [comment]
you forgot about redirects.py file, we should not 302-->404 but raise 404 right away

File: pulp_container/app/registry_api.py [hunk 4]
```diff
@@ -383,6 +384,13 @@ def create_dr(self, path, request):
 
         return distribution, repository
 
+    def handle_old_manifest_version(self, media_type, pk):
+        """
+        For non supported manifest versions, instead of redirecting the request, return earlier.
+        """
+        if media_type == MEDIA_TYPE.MANIFEST_V1:
+            raise ManifestNotFound(reference=pk)
+
```
Comment (lubosmj) [inline pulp_container/app/registry_api.py:393]: If a manifest's media type is `MEDIA_TYPE.MANIFEST_V1`, we should not raise `ManifestNotFound`. Instead, let's return it because we have got the same media type as requested by a user. Also, the `media_type` variable can contain comma separated values.

File: pulp_container/app/registry_api.py [hunk 5]
```diff
@@ -962,6 +970,9 @@ def get(self, request, path, pk):
     @RegistryApiCache(base_key=lambda req, cac: find_base_path_cached(req, cac))
     def handle_safe_method(self, request, path, pk):
         """Handles safe requests for Blobs."""
+        # Fail early in case of unsupported manifest schemas
+        self.handle_old_manifest_version(request.accepted_media_type, pk)
```
Comment (lubosmj) [inline pulp_container/app/registry_api.py:974]: We do not care about media types when it comes to blobs. You can safely remove this check here.

File: pulp_container/app/redirects.py [hunk 6]
```diff
@@ -38,6 +38,13 @@ def issue_tag_redirect(self, tag):
         """
         Issue a redirect for the passed tag.
         """
+        manifest_media_type = tag.tagged_manifest.media_type
+        if (
+            manifest_media_type not in get_accepted_media_types(self.request.headers)
+            and manifest_media_type != MEDIA_TYPE.MANIFEST_V1
+        ):
+            raise Http404("Unsupported manifest version: %s" % manifest_media_type)
```
Comment (lubosmj) [inline pulp_container/app/redirects.py:46]: Have you tried raising `ManifestNotFound` instead?

File: CHANGES/1509.removal [hunk 7]
```diff
@@ -0,0 +1,5 @@
+Removed the manifest schema conversion machinery. If the manifest is stored locally in the newer
+format and old clients request v2 schema1 manifest they will receive 404. v2 schema1 manifest is
+still going to be mirrored from remote source during sync if available and passed to the old clients
+on the request.
+This change also removed the ``pyjwkest`` dependency, since it is not used anymore.
```
Comment (lubosmj) [inline CHANGES/1509.removal:5]: I am not sure if this dependency removal is necessary to announce. @ipanova, what do you think?

File: pulp_container/app/redirects.py (hunk 6 above)
Comment (git-hyagi) [reply pulp_container/app/redirects.py:46]: hum....not yet after this change.
I'll try it!

File: CHANGES/1509.removal (hunk 7 above)
Comment (ipanova) [reply CHANGES/1509.removal:5]: I agree.

File: pulp_container/app/redirects.py [hunk 8]
```diff
@@ -34,10 +35,17 @@ class FileStorageRedirects(CommonRedirects):
...
+    def issue_tag_redirect(self, tag, pk):
         """
         Issue a redirect for the passed tag.
         """
+        manifest_media_type = tag.tagged_manifest.media_type
+        if (
+            manifest_media_type not in get_accepted_media_types(self.request.headers)
+            and manifest_media_type != MEDIA_TYPE.MANIFEST_V1
+        ):
+            raise ManifestNotFound(reference=pk)
```
Comment (lubosmj) [inline pulp_container/app/redirects.py:47]: Is it possible to use `tag.name` instead? If so, there is no need to have a an additional argument.

**Review by lubosmj — COMMENTED** [review]
Would you update the staging variant of the documentation, please?

staging_docs/user/tutorials/01-sync-and-host.md

File: pulp_container/app/redirects.py [hunk 9]
```diff
@@ -60,7 +68,8 @@ class S3StorageRedirects(CommonRedirects):
 
     def issue_tag_redirect(self, tag):
         """
-        Issue a redirect or perform a schema conversion if an accepted media type requires it.
+        Issue a redirect if an accepted media type requires it or return not found if manifest
+        version is not supported.
```
Comment (ipanova) [inline pulp_container/app/redirects.py:72]: your changes are not really related but i noticed and think that we should swap the if/elif branches

**Review by ipanova — APPROVED** [review]
looks good,thanks!

**Comment by ipanova** [comment]
@lubosmj  seems like pull-through tests is failing on azure, a fluke?

**Comment by lubosmj** [comment]
> @lubosmj seems like pull-through tests is failing on azure, a fluke? https://github.com/pulp/pulp_container/actions/runs/8111223816/job/22170105281?pr=1529#step:14:2184

I have noticed this. It started failing a couple weeks ago. It might be related to https://github.com/pulp/pulp_container/issues/1493 or artifacts/content being already present in the storage and there is a race condition. I will open an issue for that next week if it occurs again.

**Comment by ipanova** [comment]
> or artifacts/content being already present in the storage and there is a race condition. 

This is test of pull operation and not push, so no.

**Comment by lubosmj** [comment]
Yes, the error is the same as noted in the aforementioned issue:
```
    raise NotImplementedError("Redirecting to this storage is not implemented.")

```
